### PR TITLE
Header files completion, allows for non-existent compilation database.

### PR DESCRIPTION
**Pull request by ugopozo**
Hi. Since you don't usually compile your header files directly, no compilation commands are generated for them, which results, for tools like this, that can rely on compilation databases, in either basic completion support for header files or, in worst case scenarios, in no completion support at all.

This PR introduces a heuristic to try and match a header file to its correspondent .c/.cpp/.m/.mm file and get the compile commands for the later to apply to the former. This heuristic is made configurable by adding an option, `g:deoplete#sources#clang#test_extensions`, which matches known extensions that won't generate compilation commands, like ".h" and ".hpp", to extensions that do have compilation commands and are generally named the same, like ".c" or ".cpp".

If this fails, it also introduces the option `g:deoplete#sources#clang#default_file`, which may allow the user to name e.g. "main.cpp" as the fallback file to get compilation commands for and use it in files that doesn't match anything in the compilation database. It's useful if you're using some non-standard libraries (in my case, sfml), as it makes the editor able to complete the `sf::` namespace even in header files.

This also adds a check with `os.path.isdir` for the compilation database, so it doesn't fail when the directory doesn't exist.

It also removes regular files for the clang header directory search, since in some Linux distributions (in my case, Arch) there are files like "c++-analyzer" in that directory and it was getting picked up instead of the correct clang headers.

Finally, this PR fixes what seems to be typo in the regex pattern for the compilation database in the flags files, where it was checking for "comilation_database" instead of "compilation_database". With the question mark, though, it remains backwards compatible for any one who might have picked up on it and was using it with the typo.

If there's anything you'd like me to address to get this PR accepted, just say it and I'll get it fixed. Thanks for your amazing work!

### CODE REVIEW

File: rplugin/python3/deoplete/sources/deoplete_clang.py [hunk 1]
```diff
@@ -69,14 +76,14 @@ def __init__(self, vim):
             if m is not None:
                 self.completion_flags = flags[m.end():].split()
             else :
-                m = re.match(r'^comilation_database\s*=\s*', flags)
+                m = re.match(r'^comp?ilation_database\s*=\s*', flags)
```
Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_clang.py]: `?` is typo?

**Comment by zchee** [comment]
@ugopozo Thanks PR!
I will test and translate your long message(thanks!) because I'm not at good English \o/

BTW, I was commented on 79 line. It's typo?

**Comment by zchee** [comment]
oh, typo is 
> Finally, this PR fixes what seems to be typo in the regex pattern for the compilation database in the flags files, where it was checking for "comilation_database" instead of "compilation_database". With the question mark, though, it remains backwards compatible for any one who might have picked up on it and was using it with the typo.

?
sorry, more reading your comments:(

File: rplugin/python3/deoplete/sources/deoplete_clang.py (hunk 1 above)
Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_clang.py]: Sorry, it's not your typo lol

**Comment by zchee** [comment]
@ugopozo 
Now reading your pull request. not yet real testing. but would be great and LGTM.
`deoplete#sources#clang#default_file` and `deoplete#sources#clang#test_extension` feature.
and, list comprehensions before sorting of clang version, `param` fallback, and more.

also,
> If there's anything you'd like me to address to get this PR accepted, just say it and I'll get it fixed.

Yes, Please correctly fix to my typo \o/

**Comment by ugopozo** [comment]
Hi, I'm sorry for the delay, I've been busy with personal matters.

What exactly do you need me to do? The `?` on the regex just keeps it backwards compatible to people who might have used that option with the typo. Do you really want to remove it?

**Comment by zchee** [comment]
@ugopozo Hi!
>  I'm sorry for the delay, I've been busy with personal matters.
np! Thanks for back to work this.

> The ? on the regex just keeps it backward compatible to people who might have used that option with the typo. Do you really want to remove it?
I was researching git log, so it's typo are added
https://github.com/zchee/deoplete-clang/commit/3f7c10f80ccaa69027c4d4aa332e3e78297cffbb
but, I was overlooked.
and, Yes. That's option still have not in documents(README), and I was cautioned.
https://github.com/zchee/deoplete-clang/issues/31#issuecomment-223973878
So, not backward compatible but I think no problem. wrong(typo) option is more problem than it.

> What exactly do you need me to do? 
Maybe fix `?` typo only. other than that, LGTM.

Thanks!

**Comment by ugopozo** [comment]
Done!

**Comment by zchee** [comment]
@ugopozo Thanks! will check it :)

**Comment by zchee** [comment]
@ugopozo Tested. LGTM.
Thanks!!

**Pull request merged**